tests/testlease/testlease.go

```diff
@@ -0,0 +1,235 @@
+// Package testlease manages test leases in GCS bucket, as configured by environment
+// variables.
+package testlease
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/option"
+)
+
+const (
+	maxLeaseAttempts = 10
+
+	leaseAttemptInterval = 15 * time.Second
+	ownedLeaseWaitTime   = 30 * time.Second
+)
+
+var (
+	once      sync.Once
+	gcsBucket *storage.BucketHandle
+)
+
+type leaseData struct {
+	Owner   string    `json:"owned"`
+	Expires time.Time `json:"expires"`
+}
+
+func initStorageBucket(t *testing.T) *storage.BucketHandle {
+	ctx := context.Background()
+
+	bucket := os.Getenv("KOPIA_GCS_MUTEX_BUCKET")
+	if bucket == "" {
+		t.Logf("GCS mutex bucket not defined")
+		return nil
+	}
+
+	credData, err := ioutil.ReadFile(os.Getenv("KOPIA_GCS_CREDENTIALS_FILE"))
+	if err != nil {
+		t.Logf("unable to open GCS credentials file")
+		return nil
+	}
+
+	ts, err := tokenSourceFromCredentialsJSON(ctx, credData, storage.ScopeReadWrite)
+	if err != nil {
+		t.Logf("unable to get token source: %v", err)
+		return nil
+	}
+
+	cli, err := storage.NewClient(ctx, option.WithTokenSource(ts))
+	if err != nil {
+		t.Logf("unable to initialize GCS client: %v", err)
+		return nil
+	}
+
+	return cli.Bucket(bucket)
+}
+
+func getBucket(t *testing.T) *storage.BucketHandle {
+	once.Do(func() {
+		gcsBucket = initStorageBucket(t)
+	})
+
+	return gcsBucket
+}
+
+// RunWithLease executes the provided callback while holding a timed lease.
+// The lease will be released when the callback returns or will automatically expire after specified time
+// elapses.
+func RunWithLease(t *testing.T, key string, leaseDuration time.Duration, cb func()) {
+	b := getBucket(t)
+	if b == nil {
+		t.Logf("running without lease")
+		cb()
+
+		return
+	}
+
+	t.Logf("acquiring lease on GCS object %q", key)
+	oh := b.Object(key)
+	ctx := context.Background()
+
+	myOwner := uuid.New().String()
+
+	if err := tryAcquireLease(ctx, t, oh, myOwner, leaseDuration); err != nil {
+		t.Fatalf("unable to acquire lease: %v", err)
+	}
+
+	defer releaseLeaseIfStillOwned(ctx, t, oh, myOwner)
+	cb()
```

Comment:
Not requesting a change, just an observation.
There isn't anything attempting to guarantee that `cb` does not run past the lease duration. It seems that'd it'd be to the caller to choose a lease duration that matches the test timeout.
If `cb` took a `Context`, then maybe a deadline could be set on the context and pass it down. Nevertheless it'd be up to `cb` to honor the context cancellation.